lib/web/conn_upgrade.go

```diff
@@ -84,6 +95,46 @@ func (h *Handler) connectionUpgrade(w http.ResponseWriter, r *http.Request, p ht
 	return nil, nil
 }
 
+func (h *Handler) upgradeALPNWebSocket(w http.ResponseWriter, r *http.Request, upgradeHandler ConnectionHandler) (interface{}, error) {
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+		Subprotocols: []string{
+			constants.WebAPIConnUpgradeTypeALPN,
+			constants.WebAPIConnUpgradeTypeALPNPing,
+		},
+	}
+	wsConn, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		h.log.WithError(err).Debug("Failed to upgrade weboscket.")
+		return nil, trace.Wrap(err)
+	}
+	defer wsConn.Close()
+
+	logrus.Tracef("Received WebSocket upgrade with subtype: %v.", wsConn.Subprotocol())
+
+	conn := newWebSocketALPNServerConn(wsConn)
+	ctx, cancel := context.WithCancel(r.Context())
+	defer cancel()
+
+	switch wsConn.Subprotocol() {
+	case constants.WebAPIConnUpgradeTypeALPNPing:
+		// Starts native WebSocket ping for "alpn-ping".
+		go h.startPing(ctx, conn)
+	case constants.WebAPIConnUpgradeTypeALPN:
+		// Nothing to do
+	default:
+		// Just close the connection. Upgrader hijacks the connection so no
+		// point returning an error.
+		h.log.WithError(err).Debug("Unknown or empty WebSocket subprotocol.")
+		return nil, nil
+	}
+
+	if err := upgradeHandler(ctx, conn); err != nil && !utils.IsOKNetworkError(err) {
+		h.log.WithError(err).Errorf("Failed to handle %v upgrade request.", wsConn.Subprotocol())
```

Comment:
Perhaps worth repeating the fact that we don't return an error here due to the hijacker?